File: references/lib/validators/attribution.py
"""Did the summary attribute each claim to the person who actually made it?

This is the failure mode `tw` exists to avoid, and the eval could not see it.
Feeding validate_summary a summary with every quote moved to the wrong author
scored 65 -- identical to the correct one -- because the old check asked only
whether the handle and the timestamp each appeared SOMEWHERE in the source. In
a shuffled summary both tokens are present, just not together, so every bullet
counted as grounded.

Faithful attribution is a stricter claim, and a checkable one:

  1. the (handle, timestamp) PAIR must head a real source line -- not a handle
     from one tweet wearing a timestamp from another; and
  2. the bullet's content must come from THAT line, not from a different tweet
     by a different author.

Both are deterministic. Neither needs a model to judge.
"""

from __future__ import annotations

import os
import re

__all__ = [
    "CLAIM_OVERLAP_THRESHOLD",
    "attribution_faithfulness",
    "source_lines_by_author",
]

# Source lines look like `[@TechCrunch | Aug 10 14:30]: the tweet text`, which is
# the shape every model prompt tells the model to copy from.
_SOURCE_LINE_RE = re.compile(r"\[@([A-Za-z][\w]*)\s*\|\s*([^\]]+)\]\s*:?\s*(.*)")
# Bullets carry the same pair at the end: `- claim (@TechCrunch | Aug 10 14:30)`.
#
# Trailing punctuation and wrapping brackets are tolerated after the tag, and that
# is not cosmetic. The pattern used to anchor on `\)\s*$`, which made it measure
# PUNCTUATION rather than attribution:
#
#   `- claim (@Reuters | 07:10).`     full stop      -> no match
#   `- claim ((@mchen | 07:10))`      double parens  -> no match
#
# Both are correctly attributed bullets. Neither matched, so the summary counted
# ZERO tagged bullets, and every attribution check downstream was skipped --
# validate_summary's misattribution cap is gated on `total_bullets`, so a model that
# punctuates its bullets was never checked for attribution at all.
#
# Found twice, the same way, by running real models against the misattribution task:
# gemma-4-12b scored 0 "no attributed bullets" while tagging every bullet correctly
# and getting the quoted-speaker trap right, and foundation scored 0 while actually
# FAILING that trap. An instrument that returns 0 for both a right and a wrong answer
# is not measuring the thing it claims to.
_BULLET_TAG_RE = re.compile(r"\(@([A-Za-z][\w]*)\s*\|\s*([^)]+)\)[\s.,;:!)\]]*$")

_STOPWORDS = frozenset(
    """a an and are as at be by for from has have in is it its of on or that the
    to was were will with about after all also been more new now other over than
    their there these they this those we you your our but not can into out up""".split()
)

# A bullet compresses a tweet, so it shares only some of its words. Below this
# share of the bullet's own content words, it is describing something else.
CLAIM_OVERLAP_THRESHOLD = float(os.environ.get("ZTOOLS_CLAIM_OVERLAP", "0.25"))


def _content_words(text: str) -> set[str]:
    """Meaningful lowercase words, with handles and punctuation stripped out."""
    words = re.findall(r"[A-Za-z][A-Za-z0-9'-]+", text.lower())
    return {w for w in words if w not in _STOPWORDS and len(w) > 2}


def source_lines_by_author(source_text: str) -> dict[tuple[str, str], str]:
    """Map each source line's (handle, timestamp) to what that author said.

    Keyed by the PAIR rather than by handle alone: one author can post several
    times, and a bullet that borrows a sibling tweet's timestamp is exactly the
    error this is here to catch.
    """
    lines: dict[tuple[str, str], str] = {}
    for raw in (source_text or "").splitlines():
        match = _SOURCE_LINE_RE.search(raw)
        if not match:
            continue
        handle, stamp, content = match.group(1), match.group(2).strip(), match.group(3)
        lines[(handle.lower(), stamp)] = content
    return lines


def attribution_faithfulness(text: str, source_text: str) -> tuple[int, int, list[str]]:
    """(faithful, total, reasons) over bullets that carry an attribution tag.

    `total` counts only tagged bullets, so a summary with no tags at all is
    scored elsewhere (as a formatting failure) rather than being silently
    credited as 100% faithful here.
    """
    by_author = source_lines_by_author(source_text)
    faithful = total = 0
    reasons: list[str] = []

    for raw in (text or "").splitlines():
        line = raw.rstrip()
        if not line.lstrip().startswith(("-", "*")):
            continue
        tag = _BULLET_TAG_RE.search(line)
        if not tag:
            continue
        total += 1
        handle, stamp = tag.group(1).lower(), tag.group(2).strip()

        said = by_author.get((handle, stamp))
        if said is None:
            # Distinguish an invented pair from a real handle wearing a
            # borrowed timestamp: the second is the subtler error and the
            # message should say which one happened.
            if any(h == handle for h, _ in by_author):
                reasons.append(f"@{handle} did not post at {stamp}")
            else:
                reasons.append(f"@{handle} is not in the source")
            continue

        claim = _content_words(line[: tag.start()])
        if not claim:
            reasons.append(f"@{handle} bullet has no content")
            continue
        overlap = len(claim & _content_words(said)) / len(claim)
        if overlap >= CLAIM_OVERLAP_THRESHOLD:
            faithful += 1
        else:
            reasons.append(f"@{handle}'s bullet does not match what they posted")

    return faithful, total, reasons


# An untagged summary is a formatting failure, not an attribution one, and scoring
# it here would credit "produced no attributions" as "made no mistakes".
NO_TAGS_SCORE = 0
# Below this share of correctly-attributed bullets a summary is actively misleading
# rather than merely incomplete, and the message says so.
ATTRIBUTION_POOR_RATIO = 0.5


def validate_attribution(data, source_text: str = ""):
    """Score a summary on attribution faithfulness alone, as a graded ratio.

    Deliberately NOT `validate_summary`'s all-or-nothing cap. That cap is right for
    `tw` -- one wrong attribution is disqualifying for something a user will act on --
    but it makes a poor measuring instrument: every model with a single slip lands on
    the same number, so the task separates nobody. The first full-roster sweep had the
    opposite problem for the same reason -- ten of eleven models tied at 100.

    A ratio ranks. `faithful / tagged` is the fraction of bullets whose (handle,
    timestamp) pair heads a real source line AND whose content came from that line.
    """
    text = data if isinstance(data, str) else str(data or "")
    if not text.strip():
        return 0, "empty response"
    if not source_text:
        # Cannot be assessed either way. Failing here would punish a task for how it
        # was wired rather than for what the model produced -- the same mistake as
        # scoring a summary 100 for attributions nobody checked.
        return 0, "no source to check attribution against"

    faithful, total, reasons = attribution_faithfulness(text, source_text)
    if not total:
        return NO_TAGS_SCORE, "no attributed bullets (every bullet must end with (@handle | time))"

    score = round(100 * faithful / total)
    if faithful == total:
        return score, ""
    detail = "; ".join(dict.fromkeys(reasons))[:200]
    severity = "misattributed" if faithful / total < ATTRIBUTION_POOR_RATIO else "attribution slips"
    return score, f"{severity} {total - faithful}/{total}: {detail}"


__all__ += ["ATTRIBUTION_POOR_RATIO", "NO_TAGS_SCORE", "validate_attribution"]

File: rust/src/ztools/eval/validators/attribution.rs
//! Verification that summary bullet points faithfully attribute claims to their actual authors.
//!
//! Port of `lib/validators/attribution.py`.

use regex::Regex;
use serde_json::Value;
use std::collections::{HashMap, HashSet};
use std::sync::LazyLock;

pub const CLAIM_OVERLAP_THRESHOLD: f64 = 0.25;
pub const NO_TAGS_SCORE: i64 = 0;
pub const ATTRIBUTION_POOR_RATIO: f64 = 0.5;

static SOURCE_LINE_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\[@([A-Za-z][\w]*)\s*\|\s*([^\]]+)\]\s*:?\s*(.*)").unwrap());

static BULLET_TAG_RE: LazyLock<Regex> =
    LazyLock::new(|| Regex::new(r"\(@([A-Za-z][\w]*)\s*\|\s*([^)]+)\)[\s.,;:!)\]]*$").unwrap());

static WORD_RE: LazyLock<Regex> = LazyLock::new(|| Regex::new(r"[A-Za-z][A-Za-z0-9'-]+").unwrap());

static STOPWORDS: LazyLock<HashSet<&'static str>> = LazyLock::new(|| {
    [
        "a", "an", "and", "are", "as", "at", "be", "by", "for", "from", "has", "have", "in", "is",
        "it", "its", "of", "on", "or", "that", "the", "to", "was", "were", "will", "with", "about",
        "after", "all", "also", "been", "more", "new", "now", "other", "over", "than", "their",
        "there", "these", "they", "this", "those", "we", "you", "your", "our", "but", "not", "can",
        "into", "out", "up",
    ]
    .into_iter()
    .collect()
});

pub fn content_words(text: &str) -> HashSet<String> {
    WORD_RE
        .find_iter(&text.to_lowercase())
        .map(|m| m.as_str())
        .filter(|w| !STOPWORDS.contains(w) && w.len() > 2)
        .map(|w| w.to_string())
        .collect()
}

pub fn source_lines_by_author(source_text: &str) -> HashMap<(String, String), String> {
    let mut lines = HashMap::new();
    for raw in source_text.lines() {
        if let Some(caps) = SOURCE_LINE_RE.captures(raw) {
            let handle = caps.get(1).unwrap().as_str().to_lowercase();
            let stamp = caps.get(2).unwrap().as_str().trim().to_string();
            let content = caps
                .get(3)
                .map(|m| m.as_str().to_string())
                .unwrap_or_default();
            lines.insert((handle, stamp), content);
        }
    }
    lines
}

pub fn attribution_faithfulness(text: &str, source_text: &str) -> (usize, usize, Vec<String>) {
    let by_author = source_lines_by_author(source_text);
    let mut faithful = 0;
    let mut total = 0;
    let mut reasons = Vec::new();

    for raw in text.lines() {
        let line = raw.trim_end();
        let trimmed_start = line.trim_start();
        if !trimmed_start.starts_with('-') && !trimmed_start.starts_with('*') {
            continue;
        }
        let tag = match BULLET_TAG_RE.captures(line) {
            Some(c) => c,
            None => continue,
        };
        total += 1;
        let handle = tag.get(1).unwrap().as_str().to_lowercase();
        let stamp = tag.get(2).unwrap().as_str().trim().to_string();

        let said = match by_author.get(&(handle.clone(), stamp.clone())) {
            Some(s) => s,
            None => {
                if by_author.keys().any(|(h, _)| *h == handle) {
                    reasons.push(format!("@{} did not post at {}", handle, stamp));
                } else {
                    reasons.push(format!("@{} is not in the source", handle));
                }
                continue;
            }
        };

        let tag_match = tag.get(0).unwrap();
        let claim_text = &line[..tag_match.start()];
        let claim = content_words(claim_text);
        if claim.is_empty() {
            reasons.push(format!("@{} bullet has no content", handle));
            continue;
        }

        let said_words = content_words(said);
        let overlap = claim.intersection(&said_words).count() as f64 / claim.len() as f64;
        if overlap >= CLAIM_OVERLAP_THRESHOLD {
            faithful += 1;
        } else {
            reasons.push(format!(
                "@{}'s bullet does not match what they posted",
                handle
            ));
        }
    }

    (faithful, total, reasons)
}

pub fn validate_attribution(data: &Value, source_text: &str) -> (i64, String) {
    let text = match data {
        Value::String(s) => s.clone(),
        other => other.to_string(),
    };
    if text.trim().is_empty() {
        return (0, "empty response".to_string());
    }
    if source_text.is_empty() {
        return (0, "no source to check attribution against".to_string());
    }

    let (faithful, total, reasons) = attribution_faithfulness(&text, source_text);
    if total == 0 {
        return (
            NO_TAGS_SCORE,
            "no attributed bullets (every bullet must end with (@handle | time))".to_string(),
        );
    }

    let score = (100.0 * faithful as f64 / total as f64).round() as i64;
    if faithful == total {
        return (score, String::new());
    }

    let mut unique_reasons = Vec::new();
    let mut seen = HashSet::new();
    for r in reasons {
        if seen.insert(r.clone()) {
            unique_reasons.push(r);
        }
    }
    let detail_full = unique_reasons.join("; ");
    let detail = if detail_full.len() > 200 {
        &detail_full[..200]
    } else {
        &detail_full
    };
    let severity = if (faithful as f64 / total as f64) < ATTRIBUTION_POOR_RATIO {
        "misattributed"
    } else {
        "attribution slips"
    };

    (
        score,
        format!("{} {}/{}: {}", severity, total - faithful, total, detail),
    )
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn test_validate_attribution_faithful_bullets() {
        let source = "[@TechCrunch | Aug 10 14:30]: OpenAI releases new model weights\n[@Reuters | Aug 10 15:00]: Markets rally on tech earnings";
        let summary = json!("- OpenAI released weights for their latest model (@TechCrunch | Aug 10 14:30)\n- Tech earnings sparked market rallies (@Reuters | Aug 10 15:00).");
        let (score, reason) = validate_attribution(&summary, source);
        assert_eq!(score, 100);
        assert!(reason.is_empty());
    }

    #[test]
    fn test_validate_attribution_catches_swapped_author() {
        let source = "[@TechCrunch | Aug 10 14:30]: OpenAI releases new model weights\n[@Reuters | Aug 10 15:00]: Markets rally on tech earnings";
        let summary = json!("- Tech earnings sparked market rallies (@TechCrunch | Aug 10 14:30)\n- OpenAI released weights (@Reuters | Aug 10 15:00)");
        let (score, reason) = validate_attribution(&summary, source);
        assert_eq!(score, 0);
        assert!(reason.contains("misattributed 2/2"), "got: {reason}");
    }

    #[test]
    fn source_lines_without_tags_are_not_indexed() {
        let map = source_lines_by_author("a plain line with no tag\n[@alice | Jan 5]: hello world");
        assert_eq!(map.len(), 1);
        assert_eq!(
            map.get(&("alice".to_string(), "Jan 5".to_string())),
            Some(&"hello world".to_string())
        );
    }

    #[test]
    fn faithfulness_skips_unbulleted_untagged_and_empty_claims() {
        let source = "[@TechCrunch | Aug 10 14:30]: OpenAI releases new model weights";
        let text = "plain prose line\n- no attribution tag here\n- (@TechCrunch | Aug 10 14:30)\n- OpenAI releases new model weights (@TechCrunch | Aug 10 14:30)";
        let (faithful, total, reasons) = attribution_faithfulness(text, source);
        // Empty-claim tagged bullets still count toward the total.
        assert_eq!((faithful, total), (1, 2));
        assert!(
            reasons
                .iter()
                .any(|r| r.contains("@techcrunch bullet has no content")),
            "got: {reasons:?}"
        );
    }

    #[test]
    fn unknown_authors_and_wrong_timestamps_report_distinct_reasons() {
        let source = "[@TechCrunch | Aug 10 14:30]: OpenAI releases new model weights";
        let text = "- OpenAI released weights (@GhostWriter | Aug 10 14:30)\n- OpenAI released weights (@TechCrunch | Aug 11 09:00)";
        let (_, total, reasons) = attribution_faithfulness(text, source);
        assert_eq!(total, 2);
        assert!(
            reasons
                .iter()
                .any(|r| r.contains("@ghostwriter is not in the source")),
            "got: {reasons:?}"
        );
        assert!(
            reasons
                .iter()
                .any(|r| r.contains("@techcrunch did not post at Aug 11 09:00")),
            "got: {reasons:?}"
        );
    }

    #[test]
    fn validate_attribution_guard_branches() {
        let source = "[@alice | t]: some words here";

        assert_eq!(
            validate_attribution(&json!("   "), source),
            (0, "empty response".to_string())
        );
        assert_eq!(
            validate_attribution(&json!("some text"), ""),
            (0, "no source to check attribution against".to_string())
        );

        let (score, reason) = validate_attribution(&json!("- bullet one\n- bullet two"), source);
        assert_eq!(score, NO_TAGS_SCORE);
        assert!(
            reason.contains("no attributed bullets"),
            "untagged bullets must be reported: {reason}"
        );

        // Non-string payloads render through Display and find no bullets.
        let (score, reason) = validate_attribution(&json!(["not a bullet"]), source);
        assert_eq!(score, NO_TAGS_SCORE);
        assert!(reason.contains("no attributed bullets"), "got: {reason}");
    }

    #[test]
    fn half_faithful_reports_slips_not_misattribution() {
        let source = "[@TechCrunch | Aug 10 14:30]: OpenAI releases new model weights";
        let text = "- OpenAI releases new model weights (@TechCrunch | Aug 10 14:30)\n- OpenAI released weights (@TechCrunch | Aug 11 09:00)";
        let (score, reason) = validate_attribution(&json!(text), source);
        assert_eq!(score, 50);
        assert!(
            reason.starts_with("attribution slips 1/2: "),
            "got: {reason}"
        );
        assert!(
            reason.contains("did not post at Aug 11 09:00"),
            "got: {reason}"
        );
    }

    #[test]
    fn long_unique_failure_details_are_truncated_to_two_hundred_chars() {
        let source = "[@alice | Jan 01]: note one";
        let mut text = String::new();
        for i in 0..6 {
            text.push_str(&format!(
                "- filler claim number {i} (@alice | TIMESTAMP-{i:06}-PADDINGTEXT)\n"
            ));
        }
        let (score, reason) = validate_attribution(&json!(text), source);
        assert_eq!(score, 0);
        let expected_len = "misattributed 6/6: ".len() + 200;
        assert_eq!(
            reason.len(),
            expected_len,
            "detail must be cut at exactly 200 chars: len={}",
            reason.len()
        );
    }
}
